### PR TITLE
Mocha Benchmark Runner

### PR DESCRIPTION
**Pull request by CraigMacomber**
This adds an experimental package which allows mocha tests to be run as benchmarks using becnhmark.js.

Alternatively this can be though of as a tool for writing benchmarks where the benchmarks can be run a single iteration as part of the correctness test suite, but are iterated using benchmark.js when run in `--perfMode`.

Included is a MochaReporter which writes benchmark results as json files as well as human friendly tables printed to the console. This reporter ensures that failing tests are marked as such, and don't report performance data.

All perf tests authored with this tool are automatically tagged with @Benchmark as well as a more specific type which can be used to filter them, see BenchmarkType in experimental\framework\benchmark\src\Configuration.ts.

This library was originally written for benchmarking our code which uses experimental\tree, and open sourcing it as part of fluid will allow us to use it for benchmarks of experimental\tree as well.

Not included in this PR:
Any CI pipelines to run benchmarks and collect the data. We have tooling to do this using the generated JSON files, but I'm not including it at this time. If we decide such a setup is desirable and compatible with FluidFramework's CI, we can include it at a later date.

### CODE REVIEW

File: experimental/framework/benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# @fluid-experimental/benchmark
```
Comment (tylerbutler) [inline experimental/framework/benchmark/README.md:1]: Could this be in the @fluid-internal scope? Is it intended to be used outside the repo?

File: experimental/framework/benchmark/package.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
+{
+	"name": "@fluid-experimental/benchmark",
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/framework/benchmark/package.json:2]: The intention is for it to be generally useful for benchmarking anything you can put in a mocha test.

Since it uses benchmark.js its quite good at micro benchmarks (I've used it for comparing single digit nanosecond times to compare JavaScript vs native function call overhead for example), but also works fine all the way up to application integration test level performance, even async code.

Its even possible to use it with test frameworks other than mocha (we have used with with jest as well), but I'm only including a MochaReporter and runner (other frameworks would need different runners and reporters).

Its Json output is suitable for consumption by CI, and can be used with tools like PowerBI to plot performance over time, but it also has human friendly output for manual use.

Overall I think it's quite general, and expect it can meet the general benchmarking needs of FluidFramework. If it doesn't, it could be extended to do so. That way we can keep a single centralized package for consistency, since having consistent data collection and output is important for tooling/CI integration.

I'll personally be writing benchmarks for experimental tree which I'll be including in our app's internal CI's automated benchmarking, and probably port bubble bench to use it. I won't be writing other benchmarks with it, but I was thinking if it turns out to work well with experimental tree, and anyone else is interested in using it, we would promote it from experimental.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/framework/benchmark/package.json]: Oops. I missed that this file is excluded from the prettier based auto-formatting this package uses. Fixed.

File: experimental/framework/benchmark/README.md (hunk 1 above)
Comment (CraigMacomber) [reply experimental/framework/benchmark/README.md:1]: I would like to use it outside of this repo. We have quite a lot of usage of this package in a private repo (including testing our app powered by experimental/tree), and I would like to just use the copy from fluid and delete our private one.

This does remind me though: there is another approach we could take here: This package could be published separately from fluid framework (ex: in a different repo), and then used by it. It may make sense to consider that option either now, and/or when/if we decide to make it non-experimental.

Comment (tylerbutler) [reply experimental/framework/benchmark/README.md:1]: > there is another approach we could take here: This package could be published separately from fluid framework (ex: in a different repo), and then used by it.

I like that option, because it makes "ownership" clearer and conceptually this package isn't tied to anything else in the repo. That said, it's fine to commit it here for now and we can discuss that if needed in the future during the experimental --> ready transition.

**Comment by tylerbutler** [comment]
Minor suggestion: consider adding an entry to the code review assignments file (see https://github.com/microsoft/FluidFramework/wiki/PR-review-process for more info) with appropriate owners for the package.

File: examples/data-objects/client-ui-lib/package.json [hunk 3]
```diff
@@ -78,7 +78,7 @@
     "@types/assert": "^1.5.2",
     "@types/debug": "^4.1.5",
     "@types/jsdom": "^12.0.0",
-    "@types/mocha": "^5.2.5",
+    "@types/mocha": "^8.2.2",
```
Comment (vladsud) [inline examples/data-objects/client-ui-lib/package.json:81]: any chance you can do these bumps as separate PR?

File: experimental/framework/benchmark/src/Reporter.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/*
+The MIT License (MIT)
+
+Copyright (c) 2015 Robert Klep
```
Comment (vladsud) [inline experimental/framework/benchmark/src/Reporter.ts:9]: Is this available as a package, not as a source file?
Can it screw attribution? For example, when we did React Native development in Office (native apps), we would generate attribution file (i.e. all references to all copyrights / contributors) and I believe it was based on extracting licenses from packages (i.e. we have "LICENSE" file in each package).
I'm not a layer, but I think LICENSE file in this package should have attribution to the author.

**Comment by vladsud** [comment]
Given that this is not Fluid-specific, why should it be part of Fluid repo? Should it be stand-alone package?

**Comment by CraigMacomber** [comment]
> Given that this is not Fluid-specific, why should it be part of Fluid repo? Should it be stand-alone package?

Putting it in its own stand-alone package/repo is an option. I don't know the full process for doing that, so I can't do a good job of evaluating the cost/benefit for it. (I should learn how to do that though, but currently I can't evaluate the overhead very well)

I'm not sure what the bar is for something to be included in this repo. We have lots of general purpose things in `common\lib\common-utils` and `packages\test`: I think this code can serve as an improvement over `common\lib\common-utils\bench` and/or an addition to `packages\test` . I'm not sure how `common-utils\bench` is supposed to work, since it only has a dev dep on benchmark.js: I think it would need to be in its own package (so it could be depended on as a dev dep) that has a non-dev dep on benchmark.js. The package in this PR effectively provides that, along with several extra features.

In the future, I think we might want to integrate benchmarking into our CI, and I think that might be a bit easier to deal with if the package is in this repo (less FI to deal with)? I'm unsure if there is significant benefit to that if its a public package, but maybe?


Anyway, thats all the context I have for why we might want to include this in the FluidFramework repo or its own separate repo.

That leaves us with the choice (assuming we want this at all):
- Put this package in its own Repo, abandon this PR. Use that package in where desired in this repo (ex: experimental/tree).
- Go go forward with this PR into experimental, and use it in experimental/tree and bubble bench. Reevaluate where it should live in the future (ex: when we have a non-experimental user).

I'm fine with either of these.

Based on comments on this PR, I'm thinking we want to go with the first option? Anyone have a strong opinion?

File: examples/data-objects/client-ui-lib/package.json (hunk 3 above)
Comment (DLehenbauer) [reply examples/data-objects/client-ui-lib/package.json]: > Given that this is not Fluid-specific, why should it be part of Fluid repo? Should it be stand-alone package?

Craig is kindly offering to contribute his harness to Fluid to help us begin automated performance testing, which is something Curtis and I had expressed a desire for.

The fact that Craig packaged it so nicely as a stand-alone/reusable package with standardized reporting is an unexpected bonus and does open the possibility of maintaining the package separately.

I'm torn.  Part of me selfishly wants it in 'Fluid-Experimental' so it's convenient to tweak while bringing up our first few benchmarks, but it's probably a better deal for @CraigMacomber to host it independently so his commits aren't gated on Fluid developers and processes.

@CraigMacomber - Let me talk this over with @curtisman during our Tuesday sync and see if he has a desire for this to be an "in house" tool.  If we do publish it independently, I can help whip up a public GH repo with automated CI.

**Comment by tylerbutler** [comment]
> In the future, I think we might want to integrate benchmarking into our CI, and I think that might be a bit easier to deal with if the package is in this repo (less FI to deal with)?

You're probably right that a separate repo will cause some pain when you want to integrate with CI, assuming you need to make changes to both sides to get CI fully working.

> Based on comments on this PR, I'm thinking we want to go with the first option? Anyone have a strong opinion?

I wouldn't say my opinion is strong, but I do prefer keeping it independent. My most practical reason is that currently everyone has to build the whole client repo, and so every package that's added is more compilation that everyone has to do. If/when we're able to get workspaces properly set up, and people can only build the workspaces they're using, that'll be less of a concern. I can help get a repo set up if needed.

@DLehenbauer Do you have a preference?

**Comment by tylerbutler** [comment]
Setting up a new repo is admittedly more work, so I am ok with moving this PR forward and taking the discussion to another venue. There's a broader, "what goes in this repo" discussion that's happening and this is just one example.

**Comment by CraigMacomber** [comment]
@DLehenbauer Is going to take over driving this. I believe he has reasons for it to go into FluidFramework.

**Comment by DLehenbauer** [comment]
Yep.  Chatted with @curtisman and we'd like to keep the tool "close by" (at least initially) so it's easy to tweak while we're evaluating.  Curtis suggested a new "fluid-tools/" NPM namespace for non-private tools packages we publish for our own use.

**Comment by tylerbutler** [comment]
@DLehenbauer @CraigMacomber Sounds good to me.

File: experimental/framework/benchmark/src/Reporter.ts (hunk 4 above)
Comment (CraigMacomber) [reply experimental/framework/benchmark/src/Reporter.ts:9]: This file has been very heavily modified from its original form, so if you found it in its original form, that would not meet our needs anyway. I've rewritten most of it multiple times (ex: to make it much more strongly typed) and much of the original logic has been removed. It is still clearly derived from Robert Klep's original though, so I suspect its still "substantial portions of the Software"

In my prior use of this code, since it was only used as a dev dependency and never distributed other than as the full source, so I believe this setup was fine, however with it going into fluid-framework I'm not sure what the right approach is other than ask the layers.

Maybe there is something easy we can do since this code is MIT like fluid framework: perhaps add Robert Klep to the authors or something?

We have an MS internal repo showing the version history of this file as it diverged from the original: let me know if that would help here.

**Review by DLehenbauer — APPROVED** [review]
